common/types/ballot.go

```diff
@@ -0,0 +1,166 @@
+// serialized and signed to produce the signature in Ballot.
+type InnerBallot struct {
+	// the smesher's ATX in the epoch this ballot is cast.
+	AtxID ATXID
+	// the proof of the smesher's eligibility to vote and propose block content in this epoch.
+	EligibilityProof VotingEligibilityProof
+
+	// a smesher creates votes in the following steps:
+	// - select a Ballot in the past as a base Ballot
+	// - calculate the opinion difference on history between the smesher and the base Ballot
+	// - encode the opinion difference in 3 list:
+	//	 - ForDiff
+	//	   contains blocks we support while the base block did not support (i.e. voted against)
+	//	   for blocks we support in layers later than the base block, we also add them to this list
+	//   - AgainstDiff
+	//     contains blocks we vote against while the base block explicitly supported
+	//	 - NeutralDiff
+	//	   contains blocks we vote neutral while the base block explicitly supported or voted against
+	//
+	// example:
+	// layer | unified content block
+	// -----------------------------------------------------------------------------------------------
+	//   N   | UCB_A (genesis)
+	// -----------------------------------------------------------------------------------------------
+	//  N+1  | UCB_B base:UCB_A, for:[UCB_A], against:[], neutral:[]
+	// -----------------------------------------------------------------------------------------------
+	//  N+2  | UCB_C base:UCB_B, for:[UCB_B], against:[], neutral:[]
+	// -----------------------------------------------------------------------------------------------
+	//  (hare hasn't terminated for N+2)
+	//  N+3  | UCB_D base:UCB_B, for:[UCB_B], against:[], neutral:[UCB_C]
+	// -----------------------------------------------------------------------------------------------
+	//  (hare succeeded for N+2 but failed for N+3)
+	//  N+4  | UCB_E base:UCB_C, for:[UCB_C], against:[UCB_D], neutral:[]
+	// -----------------------------------------------------------------------------------------------
+	// NOTE on neutral votes: a base block is by default neutral on all blocks and layers that come after it, so
+	// there's no need to explicitly add neutral votes for more recent layers.
+	// TODO: optimize this data structure in two ways:
+	//   - neutral votes are only ever for an entire layer, never for a subset of blocks.
+	//   - collapse AgainstDiff and ForDiff into a single list.
+	//   see https://github.com/spacemeshos/go-spacemesh/issues/2369.
+	BaseBallot  BallotID
+	AgainstDiff []BlockID
+	ForDiff     []BlockID
+	NeutralDiff []BlockID
+
+	// the first Ballot the smesher cast in the epoch. this Ballot is a special Ballot that contains information
+	// that cannot be changed mid-epoch.
+	// - ActiveSet: from the smesher's view, the set of ATXs eligible to vote and propose block content in this epoch
+	// - Beacon: the beacon value the smesher recorded for this epoch
+	RefBallot BallotID
+	ActiveSet *[]ATXID
+	Beacon    []byte
```

Comment:
it will be cleaner if ActiveSet and Beacon were a part of the same struct and we should have only a pointer to that struct. also why do we define `ActiveSet *[]ATXID` as a pointer to a pointer? it makes sense only in very specific cases

Reply:
done and done